File: go/normalizer.go
package mp

import (
	"strings"
	"unicode"
)

const minLet = 2

// TextNormalizer - здесь мапы для нормализации.
type TextNormalizer struct {
	charReplacements map[string]string
	homoglyphs       map[rune]rune
	translitMap      map[string][]string
	longPatterns     map[string]string
}

// NewTextNormalizer - инициализация нормализатора.
func NewTextNormalizer(config *Config) *TextNormalizer {
	return &TextNormalizer{
		charReplacements: config.CharReplacements,
		translitMap: config.TranslitMap,
		longPatterns:     config.LongPatterns,
		homoglyphs: map[rune]rune{
			'a': 'а',
			'e': 'е',
			'o': 'о',
			'p': 'р',
			'c': 'с',
			'x': 'х',
			'y': 'у',
			'k': 'к',
			'b': 'б',
			'm': 'м',
			'h': 'н',
			'i': 'и',
			'l': 'л',
		},
	}
}

// normalize - нормализация ника.
func (n *TextNormalizer) normalize(text string) string {
    if text == "" {
        return ""
    }
    text = strings.ToLower(text)

    // Замена цифр и спецсимволов.
    for from, to := range n.charReplacements {
        text = strings.ReplaceAll(text, from, to)
    }

    // Гомоглифы.
    var builder strings.Builder
    for _, r := range text {
        if repl, ok := n.homoglyphs[r]; ok {
            builder.WriteRune(repl)
        } else {
            builder.WriteRune(r)
        }
    }
    text = builder.String()

    // Транслитерация.
    text = n.latToCyr(text)

    // Оставляем только буквы.
    clean := make([]rune, 0)
    for _, r := range text {
        if unicode.IsLetter(r) {
            clean = append(clean, r)
        }
    }

    return collapseRepeats(string(clean))
}

// latToCyr - замена в строке на русские символы.
func (n *TextNormalizer) latToCyr(text string) string {
	result := text
	for lat, cyr := range n.longPatterns {
		result = strings.ReplaceAll(result, lat, cyr)
	}
	for cyr, latVariants := range n.translitMap {
		for _, lat := range latVariants {
			result = strings.ReplaceAll(result, lat, cyr)
		}
	}
	return result
}

// collapseRepeats - убирает из строки повторения символов (Пример: ааааа).
func collapseRepeats(s string) string {
	var result []rune
	var prev rune
	count := 0

	for _, r := range s {
		if r == prev {
			count++
			if count < minLet {
				result = append(result, r)
			}
		} else {
			prev = r
			count = 1
			result = append(result, r)
		}
	}
	return string(result)
}

File: python/normalizer.py
"""
Нормализатор текста. Используется для ML-моделей.
Этапы: замена цифр/символов, гомоглифов, транслитераций, мерджим повторы.
"""

import json
from pathlib import Path

# Минимальное допустимое число одинаковых символов подряд (аналог minLet из Go).
MIN_LET = 2

# Гомоглифы: латинские буквы, визуально похожие на кириллицу.
HOMOGLYPHS: dict[str, str] = {
    'a': 'а', 'e': 'е', 'o': 'о', 'p': 'р', 'c': 'с',
    'x': 'х', 'y': 'у', 'k': 'к', 'b': 'б', 'm': 'м',
    'h': 'н', 'i': 'и', 'l': 'л',
}


class TextNormalizer:
    """
    Нормализует никнейм к единому виду для последующей проверки.
    Порядок этапов должен совпадать с Go-реализацией.
    """

    def __init__(self, config: dict):
        self.char_replacements: dict[str, str] = config.get("char_replacements", {})
        self.translit_map: dict[str, list[str]] = config.get("translit_map", {})
        self.long_patterns: dict[str, str] = config.get("long_patterns", {})

    def normalize(self, text: str) -> str:
        """Нормализует строку к кириллице без спецсимволов и повторов."""
        if not text:
            return ""
        text = text.lower()

        # Замена цифр и спецсимволов на буквы
        for src, dst in self.char_replacements.items():
            text = text.replace(src, dst)

        # Гомоглифы
        text = "".join(HOMOGLYPHS.get(ch, ch) for ch in text)
        # Транслитерация
        text = self._lat_to_cyr(text)
        # Оставляем только буквы
        text = "".join(ch for ch in text if ch.isalpha())
        # Убираем многократные повторы одной буквы
        return self._collapse_repeats(text)

    def _lat_to_cyr(self, text: str) -> str:
        """Заменяет многосимвольные паттерны (sh→ш), затем односимвольные (s→с)."""
        for lat, cyr in self.long_patterns.items():
            text = text.replace(lat, cyr)
        for cyr, lat_variants in self.translit_map.items():
            for lat in lat_variants:
                text = text.replace(lat, cyr)
        return text

    def _collapse_repeats(self, s: str) -> str:
        # Убираем повторы.
        result: list[str] = []
        prev = ""
        count = 0
        for ch in s:
            if ch == prev:
                count += 1
                if count < MIN_LET:
                    result.append(ch)
            else:
                prev = ch
                count = 1
                result.append(ch)
        return "".join(result)

def load_normalizer(config_path: str | Path) -> TextNormalizer:
    with open(config_path, encoding="utf-8") as f:
        config = json.load(f)
    return TextNormalizer(config)
